File: frontend/src/types/status.ts
import type { EntityConfig } from './domain';

export type CallState = 'planned' | 'approach' | 'moored' | 'departed';
export const ALL_CALL_STATE: readonly CallState[] = ['planned', 'approach', 'moored', 'departed'];
export type ClearanceState = 'pending' | 'cleared' | 'restricted' | 'expired';
export const ALL_CLEARANCE_STATE: readonly ClearanceState[] = ['pending', 'cleared', 'restricted', 'expired'];

export const ENTITY_CONFIGS: readonly EntityConfig[] = [
  { key: 'vesselCall', path: 'vessels', label: '船舶靠泊', statuses: ['planned', 'approach', 'moored', 'departed'] as const },
  { key: 'mooringPlan', path: 'plans', label: '系泊方案', statuses: ['draft', 'review', 'approved', 'superseded'] as const },
  { key: 'weatherWindow', path: 'weather-windows', label: '风浪窗口', statuses: ['forecast', 'safe', 'restricted', 'expired'] as const },
  { key: 'safetyClearance', path: 'clearance', label: '安全许可', statuses: ['pending', 'cleared', 'restricted', 'expired'] as const }
];

File: backend/internal/constants/status.go
package constants

// Shared status values are mirrored in frontend/src/types/status.ts. Keeping
// the lists explicit makes state-machine drift visible during code review.

type CallState string

const (
	CallStatePlanned  CallState = "planned"
	CallStateApproach CallState = "approach"
	CallStateMoored   CallState = "moored"
	CallStateDeparted CallState = "departed"
)

var AllCallState = []string{"planned", "approach", "moored", "departed"}

type ClearanceState string

const (
	ClearanceStatePending    ClearanceState = "pending"
	ClearanceStateCleared    ClearanceState = "cleared"
	ClearanceStateRestricted ClearanceState = "restricted"
	ClearanceStateExpired    ClearanceState = "expired"
)

var AllClearanceState = []string{"pending", "cleared", "restricted", "expired"}

var VesselCallTransitions = map[string]map[string]bool{
	"planned":  {"approach": true, "moored": true},
	"approach": {"moored": true, "departed": true, "planned": true},
	"moored":   {"departed": true, "approach": true},
	"departed": {"moored": true},
}

var MooringPlanTransitions = map[string]map[string]bool{
	"draft":      {"review": true, "approved": true},
	"review":     {"approved": true, "superseded": true, "draft": true},
	"approved":   {"superseded": true, "review": true},
	"superseded": {"approved": true},
}

var WeatherWindowTransitions = map[string]map[string]bool{
	"forecast":   {"safe": true, "restricted": true},
	"safe":       {"restricted": true, "expired": true, "forecast": true},
	"restricted": {"expired": true, "safe": true},
	"expired":    {"restricted": true},
}

var SafetyClearanceTransitions = map[string]map[string]bool{
	"pending":    {"cleared": true, "restricted": true},
	"cleared":    {"restricted": true, "expired": true, "pending": true},
	"restricted": {"expired": true, "cleared": true},
	"expired":    {"restricted": true},
}

func CanTransition(graph map[string]map[string]bool, from, to string) bool {
	targets, exists := graph[from]
	return exists && targets[to]
}
